Fix interprocessor interrupts in gic-v2

razvanvirtan opened the pull request:
This PR fixes the SGI generation and interrupts ack functionalities on gic-v2.
A deadlock in the SGI implementation is fixed in the already existing implementation. Also, SGIs are added to our GIC generic interface.

Signed-off-by: Răzvan Vîrtan <virtanrazvan@gmail.com>

rene commented on plat/drivers/gic/gic-v3.c:
```diff
@@ -606,3 +606,43 @@ int gicv3_probe(const void *fdt, struct _gic_dev **dev)
+    sgi_register |= (sgintid << 24);
+
+	/* Set affinity fields and optional range selector */
+	sgi_register |= (cpuid & MPIDR_AFF3_MASK) << 16;
```
Shift bits are not in accordance with affinity positions in ICC_SGI_ELn register: 
https://developer.arm.com/documentation/ddi0595/2020-12/AArch64-Registers/ICC-SGI1R-EL1--Interrupt-Controller-Software-Generated-Interrupt-Group-1-Register

razvanvirtan replied:
Done, sorry for this, I don't see how I missed it.

rene commented on plat/drivers/include/gic/gic-v3.h:
```diff
@@ -374,4 +374,12 @@
+ * @sigintid the software generated interrupt id
+ * @cpuid the id of the targeted cpu
+ */
+void gic_sgi_gen(uint8_t sigintid, uint32_t cpuid);
```
I think the best approach for GIC driver is actually add gic_sgi_gen function to struct _gic_operations, then set corresponding static function for both GICv2 and GICv3.

razvanvirtan replied:
I've updated the operations interface.

rene commented on plat/drivers/gic/gic-v2.c:
```diff
@@ -179,11 +179,17 @@ void gic_sgi_gen_to_list(uint32_t sgintid, uint8_t targetlist)
 	unsigned long irqf;
 
 	/* spin lock here is needed when smp is supported */
-	dist_lock(gicv2_drv);
```
@razvanvirtan , please, break your changes here into separated commits. I would suggest one commit for the deadlock fix, another commit to add the SGI interface to the generic GIC driver struct and eventually a third commit to the SGI generation implementation for GICv3.....

razvanvirtan replied:
@rene Done. The problematic commit is the third one, SGIs don't reach their destination when using gic-v3. You can see this when running the test for my SMP pull request https://github.com/unikraft/unikraft/pull/373, the application hangs forever because the `unlock()` in the interrupt handler is never reached.

rene replied:
@razvanvirtan looks good to me. According to your last comment, so basically SGI generation is not working on gic-v3? I will give a try in your SMP branch...

razvanvirtan replied:
Yes, when using gic-v3 with my SMP bootcode the application blocks.

razvanvirtan replied:
Also, note that you will have to select uklock when testing the SMP draft, I've added this in the SMP PR too (I think I will update the PR so that uklock is included by default when SMP is enabled).

razvanvirtan committed:
plat/drivers/gic: Fix deadlock in GIC-v2 SGI

At this point, SGIs can't be generated using gic-v2
because a deadlock occurs. This commit repairs the deadlock,
removing a redundant lock operation.

Signed-off-by: Răzvan Vîrtan <virtanrazvan@gmail.com>

razvanvirtan committed:
plat/drivers/gic: Add SGI generation to the gic interface

Inter-processor interrupts support is implemented in gic-v2,
but the gic interface and the interrupt controller don't
expose it. This commit adds SGI to the common interrupt controller
interface.

Signed-off-by: Răzvan Vîrtan <virtanrazvan@gmail.com>

michpappas commented:
@rene do you have any further comments on this PR? If not could you please add your approval so that we can merge it in v0.9?

rene commented:
@michpappas , well, as pointed out by @razvanvirtan, the SGI generation for GICv3 doesn't work with SMP (there is a deadlock), I would prefer to wait for SMP PR before merge this one.... I will take a look on this one with SMP to try to figure out where the issue is (on SMP code or on GIC code)....

michpappas commented:
@rene, @razvanvirtan there is a circular dependency here 🐔 🥚 Without SGI support we cannot merge the SMP PR. @razvanvirtan would it make sense to break this down into two separate PRs, one with the changes required in GICv2 that we can merge now, and another one for GICv3 that @rene can have a look at and we merge in the next release?

rene commented:
@michpappas does it make sense, @razvanvirtan can purge from this PR only SGI GICv3 implementation, all other commits are fine and reviewed (from my side)....

michpappas commented:
@rene yes that's what I meant earlier, but maybe didn't make it clear, thanks :)

michpappas commented:
@razvanvirtan, @rene my understanding here is that the deadlock was due to locking the distributor both here and also in `gic_sgi_gen()`. Is that correct? If that is the case, then I see the same problem in `gic_sgi_gen_to_others()`.

There are also these comments in both `gic_sgi_gen_to_list()` and `gic_sgi_gen_to_others()`:
```
/* spin lock here is needed when smp is supported */
```
@rene since you previously worked on SMP support on GIC, can you confirm if these are valid?

In the two same functions, this comment can also be removed.
```
TODO: this will not work until SMP is supported
```

razvanvirtan commented:
> @razvanvirtan, @rene my understanding here is that the deadlock was due to locking the distributor both here and also in `gic_sgi_gen()`. Is that correct? If that is the case, then I see the same problem in `gic_sgi_gen_to_others()`.
> 
> There are also these comments in both `gic_sgi_gen_to_list()` and `gic_sgi_gen_to_others()`:
> 
> ```
> /* spin lock here is needed when smp is supported */
> ```
> 
> @rene since you previously worked on SMP support on GIC, can you confirm if these are valid?
> 
> In the two same functions, this comment can also be removed.
> 
> ```
> TODO: this will not work until SMP is supported
> ```

@michpappas Indeed, the same issue occured in `gic_sgi_gen_to_others()` too,  I have fixed it. Also moved the spin lock comment in `gic_sgi_gen()` and removed the TODOs.

michpappas commented:
@razvanvirtan thanks 🙂
@rene, or someone from @unikraft/maintainers-core can you please add your RB so that I can merge?

rene reviewed (approved):
@razvand , thanks for the update, looks good now.

Reviewed-by: Renê de Souza Pinto <rene@renesp.com.br>

rene commented:
> > @razvanvirtan, @rene my understanding here is that the deadlock was due to locking the distributor both here and also in `gic_sgi_gen()`. Is that correct? If that is the case, then I see the same problem in `gic_sgi_gen_to_others()`.
> > There are also these comments in both `gic_sgi_gen_to_list()` and `gic_sgi_gen_to_others()`:
> > ```
> > /* spin lock here is needed when smp is supported */
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @rene since you previously worked on SMP support on GIC, can you confirm if these are valid?
> > In the two same functions, this comment can also be removed.
> > ```
> > TODO: this will not work until SMP is supported
> > ```
> 
> @michpappas Indeed, the same issue occured in `gic_sgi_gen_to_others()` too, I have fixed it. Also moved the spin lock comment in `gic_sgi_gen()` and removed the TODOs.

That's correct. I've reviewed your changes, all good from my side.

michpappas commented:
@rene thanks a lot!

Approved-by: Michalis Pappas <michalis.pappas@opensynergy.com>
